### PR TITLE
Improve OData query handling

### PR DESCRIPTION
**Pull request by SpeckiJ**
- Various bugfixes
- Test parsing for all permutations of input query string
- Implement Stringification of Expr to directly output as OData conform query string

### CODE REVIEW

**Comment by autermann** [comment]
`CI=true` oder `-Pcheck` 😉

**Comment by SpeckiJ** [comment]
I could swear that spotbugs did not find anything on my machine...

@autermann Is there a way to trigger a travis rebuild? Just push an empty commit? The last CI failed because of internal issues with Travis (i guess?)

**Comment by autermann** [comment]
You can do that in the Travis Website in the view for the particular build, but I'm not sure if you have the rights to do that... But I've triggered a rebuild.

File: shetland/core/src/main/java/org/n52/shetland/filter/CountFilter.java [hunk 1]
```diff
@@ -53,4 +53,7 @@ public Boolean getValue() {
         return this.value.equals(((CountFilter) o).getValue());
     }
 
+    @Override public String toString() {
+        return DOLLAR + "count=" + getValue().toString();
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/CountFilter.java:57]: `return "$count=" + getValue();`

File: shetland/core/src/main/java/org/n52/shetland/filter/ExpandFilter.java [hunk 2]
```diff
@@ -55,4 +56,9 @@ public ExpandFilter(Set<ExpandItem> items) {
 
         return this.items.equals(((ExpandFilter) o).getItems());
     }
+
+    @Override public String toString() {
+        return DOLLAR + "expand=" +
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/ExpandFilter.java:61]: `return "$expand=" +`

File: shetland/core/src/main/java/org/n52/shetland/filter/ExpandItem.java [hunk 3]
```diff
@@ -62,4 +62,8 @@ public QueryOptions getQueryOptions() {
         return Objects.equals(this.path, ((ExpandItem) o).getPath()) &&
                 Objects.equals(this.queryOptions, ((ExpandItem) o).getQueryOptions());
     }
+
+    @Override public String toString() {
+        return path + "(" + getQueryOptions().toString() + ")";
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/ExpandItem.java:67]: `return path + "(" + getQueryOptions() + ")";`

File: shetland/core/src/main/java/org/n52/shetland/filter/FilterFilter.java [hunk 4]
```diff
@@ -55,4 +54,8 @@ public Object getFilter() {
 
         return Objects.equals(((FilterFilter) o).getFilter(), this.getFilter());
     }
+
+    @Override public String toString() {
+        return DOLLAR + "filter=" + filter.toODataString();
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/FilterFilter.java:59]: `return "$filter=" + filter.toODataString();`

File: shetland/core/src/main/java/org/n52/shetland/filter/OrderByFilter.java [hunk 5]
```diff
@@ -71,4 +72,8 @@ public OrderByFilter(List<OrderProperty> sortProperties) {
         return this.sortProperties.equals(((OrderByFilter) o).getSortProperties());
     }
 
+    @Override public String toString() {
+        return DOLLAR + "orderby=" +
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/OrderByFilter.java:76]: `return "$orderby=" +`

File: shetland/core/src/main/java/org/n52/shetland/filter/SelectFilter.java [hunk 6]
```diff
@@ -55,4 +55,8 @@ public SelectFilter(Set<String> items) {
 
         return Objects.equals(this.getItems(), ((SelectFilter) o).getItems());
     }
+
+    @Override public String toString() {
+        return DOLLAR + "select=" + String.join(",", this.items);
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/SelectFilter.java:60]: `return "$select=" + String.join(",", this.items);`

File: shetland/core/src/main/java/org/n52/shetland/filter/SkipTopFilter.java [hunk 7]
```diff
@@ -59,4 +59,8 @@ public Long getValue() {
         return Objects.equals(this.operator, ((SkipTopFilter) o).getOperator())
                 && Objects.equals(this.value, ((SkipTopFilter) o).getValue());
     }
+
+    @Override public String toString() {
+        return DOLLAR + ((getOperator().equals(SkipTopOperator.Top)) ? "top" : "skip") + "=" + getValue().toString();
```
Comment (autermann) [inline shetland/core/src/main/java/org/n52/shetland/filter/SkipTopFilter.java:64]: `return "$" + ((getOperator().equals(SkipTopOperator.Top)) ? "top" : "skip") + "=" + getValue();`

File: svalbard/odata/src/main/java/org/n52/svalbard/odata/core/expr/GeoValueExpr.java [hunk 8]
```diff
@@ -60,6 +60,10 @@ public String toString() {
         return String.format("'%s'", this.value);
     }
 
+    @Override public String toODataString() {
+        return String.format("%s", this.value);
```
Comment (autermann) [inline svalbard/odata/src/main/java/org/n52/svalbard/odata/core/expr/GeoValueExpr.java:64]: `return this.value`;

File: svalbard/odata/src/main/java/org/n52/svalbard/odata/core/expr/arithmetic/NumericValueExpr.java [hunk 9]
```diff
@@ -73,6 +73,10 @@ public String toString() {
         return String.format("As float: '%f'", this.value.floatValue());
     }
 
+    @Override public String toODataString() {
+        return String.format("%f", this.value.floatValue());
```
Comment (autermann) [inline svalbard/odata/src/main/java/org/n52/svalbard/odata/core/expr/arithmetic/NumericValueExpr.java:77]: `return Float.toString(this.value.floatValue());`

File: shetland/core/src/main/java/org/n52/shetland/oasis/odata/query/option/QueryOptions.java [hunk 10]
```diff
@@ -171,4 +198,39 @@ public boolean equals(Object o) {
...
+        if (this.hasOrderByFilter()) {
+            builder.append(this.getOrderByFilter().toString());
+            builder.append(AMPERSAND);
+        }
+        if (this.hasFilterFilter()) {
+            builder.append(this.getFilterFilter().toString());
+            builder.append(AMPERSAND);
+        }
+        return (builder.length() > 0) ? builder.substring(0, builder.length() - 1) : "";
+    }
```
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply shetland/core/src/main/java/org/n52/shetland/oasis/odata/query/option/QueryOptions.java:235]: Definitely much cleaner. will change.
`getAllFilters` did not exist when i initially wrote this Method.

**Pull request merged**